mqtt/mqtt-bridge/src/client.rs

```diff
@@ -32,55 +40,157 @@ impl ShutdownHandle {
-        let token_source = self.token_source.as_ref().cloned();
+        match self {
+            BridgeIoSource::Tcp(connect_settings) => Self::get_tcp_source(connect_settings.clone()),
+            BridgeIoSource::Tls(connect_settings) => Self::get_tls_source(&connect_settings),
+        }
+    }
+}
+
+impl BridgeIoSource {
+    fn get_tcp_source(connection_settings: TcpConnection<SasTokenSource>) -> BridgeIoSourceFuture {
+        let address = connection_settings.address;
+        let token_source = connection_settings.token_source;
+
+        Box::pin(async move {
+            let expiry = Utc::now() + chrono::Duration::minutes(DEFAULT_TOKEN_DURATION_MINS);
+
+            let io = TcpStream::connect(&address);
+
+            let token_task = async {
+                match token_source {
+                    Some(ref ts) => ts.get(&expiry).await,
+                    None => Ok(None),
+                }
+            };
+
+            let (password, io) = future::try_join(token_task, io).await.map_err(|err| {
+                Error::new(ErrorKind::Other, format!("failed to connect: {}", err))
+            })?;
+
+            let stream: Pin<Box<dyn BridgeIo>> = Box::pin(io);
+            Ok((stream, password))
+        })
+    }
+
+    fn get_tls_source(connection_settings: &TcpConnection<SasTokenSource>) -> BridgeIoSourceFuture {
+        let address = connection_settings.address.clone();
+        let token_source = connection_settings.token_source.as_ref().cloned();
+        let trust_bundle_source = connection_settings.trust_bundle_source.clone();
 
         Box::pin(async move {
             let expiry = Utc::now() + chrono::Duration::minutes(DEFAULT_TOKEN_DURATION_MINS);
 
-            let password: Option<String> = if let Some(ts) = token_source {
-                ts.get(&expiry)
+            let server_root_certificate_task = async {
+                match trust_bundle_source {
+                    Some(ref source) => source.get_trust_bundle().await,
+                    None => Ok(None),
+                }
+            };
+
+            let token_task = async {
+                match token_source {
+                    Some(ref ts) => ts.get(&expiry).await,
+                    None => Ok(None),
+                }
+            };
+
+            let io = TcpStream::connect(address.clone());
+
+            let (server_root_certificate, password, stream) =
+                try_join3(server_root_certificate_task, token_task, io)
                     .await
-                    .map_err(|e| {
-                        error!("Failed to get token for connection {} {}", address, e);
-                        e
-                    })
-                    .ok()
+                    .map_err(|err| {
+                        Error::new(ErrorKind::Other, format!("failed to connect: {}", err))
+                    })?;
+
+            let config = SslConnector::builder(SslMethod::tls())
+                .map(|mut builder| {
```

Comment:
it is a bit too many combinators I guess...

Reply:
SHouldn't we return an error instead of silently converting it to Option?

Reply:
It would in the end return the error because None would result in an error, but anyway I changed it to map_err here and map_err again at line 180. I though many combinators is the  dream.